libvast/src/http_api.cpp

```diff
@@ -21,64 +21,150 @@ auto rest_endpoint::canonical_path() const -> std::string {
+    case simdjson::ondemand::json_type::number:
+    case simdjson::ondemand::json_type::boolean:
+      return std::string{value.raw_json_token()};
+    case simdjson::ondemand::json_type::string:
+      return std::string{value.get_string().value()};
+    case simdjson::ondemand::json_type::array: {
+      auto lst = value.get_array();
+      list xs;
+      for (auto x : lst)
+        xs.push_back(parse_skeleton(x.value(), depth + 1));
+      return xs;
+    }
+    case simdjson::ondemand::json_type::object: {
+      record xs;
+      auto obj = value.get_object();
+      for (auto pair : obj)
+        xs.emplace(std::string{pair.unescaped_key().value()},
+                   parse_skeleton(pair.value().value(), depth + 1));
+      return xs;
+    }
+  }
+  die("missing return in switch statement");
+}
+
+auto http_parameter_map::from_json(std::string_view json)
+  -> caf::expected<http_parameter_map> {
+  http_parameter_map result;
+  if (json.empty())
+    return result;
+  auto padded_string = simdjson::padded_string{json};
+  auto parser = simdjson::ondemand::parser{};
+  auto doc = simdjson::ondemand::document{};
+  auto body = simdjson::ondemand::object{};
+  if ([[maybe_unused]] auto error = parser.iterate(padded_string).get(doc)) {
+    // std::stringstream ss;
+    // ss << error;
+    // auto detailed_error = ss.str();
+    return caf::make_error(ec::invalid_argument, "failed to parse json");
+  }
+  if ([[maybe_unused]] auto error = doc.get_object().get(body))
+    return caf::make_error(ec::invalid_argument, "expected a top-level object");
+  for (auto obj : body)
+    result.params_.emplace(std::string{obj.unescaped_key().value()},
+                           parse_skeleton(obj.value()));
+  return result;
+}
+
+auto http_parameter_map::params() const
+  -> const detail::stable_map<std::string, vast::data>& {
+  return params_;
+}
+
+auto http_parameter_map::get_unsafe()
+  -> detail::stable_map<std::string, vast::data>& {
+  return params_;
+}
+
+auto http_parameter_map::emplace(std::string&& key, vast::data&& value)
+  -> void {
+  // TODO: Validate that `value` contains only strings, lists and records.
+  params_.emplace(std::move(key), std::move(value));
+}
+
+auto parse_endpoint_parameters(const vast::rest_endpoint& endpoint,
+                               const http_parameter_map& parameter_map)
   -> caf::expected<vast::record> {
   vast::record result;
   if (!endpoint.params)
     return result;
+  auto const& params = parameter_map.params();
   for (auto const& leaf : endpoint.params->leaves()) {
     auto const& name = leaf.field.name;
     auto maybe_param = params.find(name);
     if (maybe_param == params.end())
       continue;
-    auto const& string_value = maybe_param->second;
+    auto const& param_data = maybe_param->second;
+    auto is_string = caf::holds_alternative<std::string>(param_data);
+    auto is_list = caf::holds_alternative<vast::list>(param_data);
     auto typed_value = caf::visit(
```

Comment:
Lets add a `TODO` to signal that we want to allow more possible endpoint signatures and use a recursive algorithm here.